### PR TITLE
SNOW-736355: [HTAP] Add Retry Context to retried request

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-jy**
### Description
Please explain the changes you made here.
- [x] Added 'includeRetryReason' option.
- [x] Replaced 'retry' parameter with Retrycount, and added RetryReason when the option is enabled.
- [x] Added connection testing code.  

### Checklist
- [x] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [x] Create tests which fail without the change (if possible)
- [x] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [x] Extend the README / documentation and ensure is properly displayed (if necessary)
- [x] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

File: lib/connection/statement.js [hunk 1]
```diff
@@ -1570,7 +1572,11 @@ function sendSfRequest(statementContext, options, appendQueryParamOnRetry)
     // retry, update the url
     if ((numRetries > 0) && appendQueryParamOnRetry)
     {
-      options.url = Util.url.appendParam(urlOrig, 'retry', true);
+      let modifiedURL;
+      modifiedURL = Util.url.appendParam(urlOrig, 'retryCount', numRetries);
+      if(connectionConfig.getIncludeRetryContext){
+        modifiedURL = Util.url.appendParam(modifiedURL, 'retryReason', lastStatusCodeForRetry);
```
Comment (sfc-gh-pmotacki) [inline lib/connection/statement.js:1578]: Could you implement some tests checking retry parameters?

**Review by sfc-gh-pmotacki — CHANGES_REQUESTED** [review]

File: lib/errors.js [hunk 2]
```diff
@@ -69,6 +69,7 @@ codes.ERR_CONN_CREATE_INVALID_ARRAY_BINDING_THRESHOLD = 404037;
 codes.ERR_CONN_CREATE_INVALID_GCS_USE_DOWNSCOPED_CREDENTIAL = 404038;
 codes.ERR_CONN_CREATE_INVALID_FORCE_STAGE_BIND_ERROR = 404039;
 codes.ERR_CONN_CREATE_INVALID_BROWSER_TIMEOUT = 404040;
+codes.ERR_CONN_CREATE_INVALID_INCLUDE_RETRY_REASON =404042
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-jy) [reply lib/errors.js:72]: 404041 is for disableQueryContextCache. It is in a different PR.

File: lib/constants/error_messages.js [hunk 3]
```diff
@@ -64,6 +64,7 @@ exports[404037] = 'Invalid arrayBindingThreshold. The specified value must be a
 exports[404038] = 'Invalid gcsUseDownscopedCredential. The specified value must be a boolean.';
 exports[404039] = 'Invalid forceStageBindError. The specified value must be a number.';
 exports[404040] = 'Invalid browser timeout value. The specified value must be a positive number.';
+exports[404042] = 'Invalid Include retry reason. The specified value must be a boolean.'
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-jy) [reply lib/constants/error_messages.js:67]: Fixed

File: lib/connection/statement.js (hunk 1 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/connection/statement.js:1578]: I am trying to create a testing function. Actually, Util.url.appendParam is already tested in url_util.js, and I am unsure that testing the same function again for the retry is reasonable. Also, I think we do not have a way how we can test this with an integration test or an E2E test. Please let me know if you have any idea for this.

Comment (sfc-gh-ext-simba-jy) [reply lib/connection/statement.js]: I created a new function for this, and created a unit testing.

File: lib/services/sf.js [hunk 4]
```diff
@@ -558,6 +558,7 @@ function StateAbstract(options)
    */
   function sendHttpRequest(requestOptions, httpClient)
   {
+
```
Comment (sfc-gh-ext-simba-lf) [inline lib/services/sf.js:561]: nit: remove added line

File: lib/services/sf.js [hunk 5]
```diff
@@ -672,6 +673,7 @@ function StateAbstract(options)
       realRequestOptions.url +=
         ('clientStartTime=' + requestOptions.startTime
           + "&" + 'retryCount=' + (requestOptions.retry - 1));
+
```
Comment (sfc-gh-ext-simba-lf) [inline lib/services/sf.js:676]: nit: remove added line

File: lib/services/sf.js (hunk 4 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/services/sf.js]: Removed.

File: lib/services/sf.js (hunk 5 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/services/sf.js]: Removed.

File: lib/connection/connection_config.js [hunk 6]
```diff
@@ -484,6 +485,13 @@ function ConnectionConfig(options, validateCredentials, qaMode, clientInfo)
     }
   }
 
+  let includeRetryReason = true;
+  if (Util.exists(options.includeRetryReason)) {
+    Errors.checkArgumentValid(Util.isBoolean(options.includeRetryReason),
+      ErrorCodes.ERR_CONN_CREATE_INVALID_INCLUDE_RETRY_REASON);
+      includeRetryReason = options.includeRetryReason;
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection_config.js:492]: this line does not seems to be properly indented

File: lib/connection/connection_config.js [hunk 7]
```diff
@@ -747,6 +755,16 @@ function ConnectionConfig(options, validateCredentials, qaMode, clientInfo)
     return forceStageBindError;
   };
 
+  /**
+   * Returns whether the Retry reason is included or not in the retry url
+   *    
+   * @returns {string}
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection_config.js:761]: should it be boolean, not string?

File: test/unit/util_test.js [hunk 8]
```diff
@@ -488,6 +488,37 @@ describe('Util', function ()
     }
   });
 
+  it('Util.url.appendRetryParam()', function () {
```
Comment (sfc-gh-pmotacki) [inline test/unit/util_test.js:491]: Can you propose better title of this test?

File: lib/connection/statement.js [hunk 9]
```diff
@@ -1570,7 +1571,14 @@ function sendSfRequest(statementContext, options, appendQueryParamOnRetry)
     // retry, update the url
     if ((numRetries > 0) && appendQueryParamOnRetry)
     {
-      options.url = Util.url.appendParam(urlOrig, 'retry', true);
+      const retryOption = {
+        url: urlOrig,
+        retryCount: numRetries,
+        retryReason: lastStatusCodeForRetry,
+        includeRetryContext: connectionConfig.getIncludeRetryContext()
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/statement.js:1578]: shouldn't it be `getIncludeRetryReason`, not context?

File: lib/util.js [hunk 10]
```diff
@@ -344,6 +344,14 @@ exports.url =
       }
 
       return url;
+    },
+
+    appendRetryParam: function (option) {
+      let retryUrl = this.appendParam(option.url, 'retryCount', option.retryCount);
+      if (option.includeRetryContext) {
```
Comment (sfc-gh-dprzybysz) [inline lib/util.js:351]: is it reason related only and that how should it be named, not context?

**Review by sfc-gh-pmotacki — CHANGES_REQUESTED** [review]

File: lib/constants/error_messages.js [hunk 11]
```diff
@@ -64,6 +64,7 @@ exports[404037] = 'Invalid arrayBindingThreshold. The specified value must be a
 exports[404038] = 'Invalid gcsUseDownscopedCredential. The specified value must be a boolean.';
 exports[404039] = 'Invalid forceStageBindError. The specified value must be a number.';
 exports[404040] = 'Invalid browser timeout value. The specified value must be a positive number.';
+exports[404042] = 'Invalid include retry reason. The specified value must be a boolean.'
```
Comment (sfc-gh-pmotacki) [inline lib/constants/error_messages.js:67]: Why 404041 number was omitted?

File: test/unit/util_test.js (hunk 8 above)
Comment (sfc-gh-dprzybysz) [inline test/unit/util_test.js:491]: let's write this test with separated `it` functions:
```
[
        {
          option: {
            url: 'http://www.something.snowflakecomputing.com',
            retryCount: 3,
            retryReason: 429,
            includeRetryContext: true,
          },
          result: 'http://www.something.snowflakecomputing.com?retryCount=3&retryReason=429'
        },
        {
          option: {
            url: 'http://www.something.snowflakecomputing.com',
            retryCount: 3,
            retryReason: 429,
            includeRetryContext: false,
          },
          result: 'http://www.something.snowflakecomputing.com?retryCount=3'
        }
      ].forEach((...) => {
  it(..)
})
```
and let's it contains more description of the case e.g.
`append retry params with retry reason` and `append retry params without retry reason`

**Review by sfc-gh-pmotacki — CHANGES_REQUESTED** [review]

File: test/unit/util_test.js [hunk 12]
```diff
@@ -488,6 +488,37 @@ describe('Util', function ()
...
+            retryReason: 429,
+            includeRetryContext: false,
+          },
+          result: 'http://www.something.snowflakecomputing.com?retryCount=3'
+        }
+      ];
+
+      for (let i = 0; i < testCases.length; i++) {
+        testCase = testCases[i];
+        url = Util.url.appendRetryParam(testCase.option);
```
Comment (sfc-gh-dprzybysz) [inline test/unit/util_test.js:517]: url could be const here?

File: lib/connection/connection_config.js [hunk 13]
```diff
@@ -48,7 +48,8 @@ const DEFAULT_PARAMS =
   'validateDefaultParameters',
   'arrayBindingThreshold',
   'gcsUseDownscopedCredential',
-  'forceStageBindError'
+  'forceStageBindError',
+  'includeRetryReason'
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection_config.js:52]: let's add trailing coma

File: lib/connection/connection_config.js (hunk 7 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/connection/connection_config.js:761]: Fixed

File: lib/connection/connection_config.js (hunk 13 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/connection/connection_config.js:52]: added

File: lib/connection/connection_config.js (hunk 6 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/connection/connection_config.js:492]: added

File: lib/connection/statement.js (hunk 9 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/connection/statement.js:1578]: Fixed.

Comment (sfc-gh-ext-simba-jy) [reply lib/connection/statement.js:1578]: Fixed.

File: lib/util.js (hunk 10 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/util.js:351]: No, I was confused with the query context cache.

File: lib/connection/connection_config.js (hunk 13 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/connection/connection_config.js:52]: Added.

File: test/unit/util_test.js (hunk 12 above)
Comment (sfc-gh-ext-simba-jy) [reply test/unit/util_test.js:517]: fixed.

File: test/unit/util_test.js (hunk 8 above)
Comment (sfc-gh-ext-simba-jy) [reply test/unit/util_test.js:491]: Added.

Comment (sfc-gh-ext-simba-jy) [reply test/unit/util_test.js:491]: Changed.

File: lib/connection/connection_config.js (hunk 6 above)
Comment (sfc-gh-dprzybysz) [reply lib/connection/connection_config.js:493]: i don't think it is fixed

**Review by sfc-gh-dprzybysz — APPROVED** [review]

File: lib/connection/connection_config.js (hunk 6 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/connection/connection_config.js:493]: Interesting... I can see there is an indent between includRetryReason and ErrorCode check in the file changes.

Comment (sfc-gh-ext-simba-jy) [reply lib/connection/connection_config.js:493]: Double checked, and the space is applied.

Comment (sfc-gh-ext-simba-jy) [reply lib/connection/connection_config.js:493]: I double checked and there is a space

Comment (sfc-gh-dprzybysz) [reply lib/connection/connection_config.js:493]: when I said indent I meant not an empty line but the amount of spaces from the left side :)

Comment (sfc-gh-ext-simba-jy) [reply lib/connection/connection_config.js]: Ahhh, I see. I edited it :).

**Review by sfc-gh-pmotacki — APPROVED** [review]
OK

**Pull request merged**